test/translator/inquiry_test.cc

```diff
@@ -0,0 +1,328 @@
+  }
+}
+
+TEST_F(InquiryTest, TranslateUnitSerialNumberVpdNguid) {
+  inquiry_cmd_ = scsi::InquiryCommand{
+      .evpd = 1, .page_code = scsi::PageCode::kUnitSerialNumber};
+
+  identify_ns.eui64 = 0;
+  identify_ns.nguid[0] = 0x123456789abcdefa;
+  identify_ns.nguid[1] = 0x123456789abcdefa;
+
+  translator::StatusCode status =
+      translator::InquiryToScsi(scsi_cmd_, buffer, nvme_cmds_);
+  EXPECT_EQ(status, translator::StatusCode::kSuccess);
+
+  scsi::UnitSerialNumber result =
+      *reinterpret_cast<scsi::UnitSerialNumber*>(buffer);
+
+  EXPECT_EQ(result.peripheral_qualifier,
+            scsi::PeripheralQualifier::kPeripheralDeviceConnected);
+  EXPECT_EQ(result.peripheral_device_type,
+            scsi::PeripheralDeviceType::kDirectAccessBlock);
+  EXPECT_EQ(result.page_code, scsi::PageCode::kUnitSerialNumber);
+  EXPECT_EQ(result.page_length, 40);
+  char formatted_hex_string[41] = "1234_5678_9abc_defa_1234_5678_9abc_defa.";
+  uint8_t* product_serial_number =
+      reinterpret_cast<uint8_t*>(&buffer[sizeof(scsi::UnitSerialNumber)]);
+
+  for (int i = 0; i < 40; ++i) {
+    EXPECT_EQ(product_serial_number[i], formatted_hex_string[i]);
+  }
+}
+
+TEST_F(InquiryTest, TranslateUnitSerialNumberVpdBoth) {
+  inquiry_cmd_ = scsi::InquiryCommand{
+      .evpd = 1, .page_code = scsi::PageCode::kUnitSerialNumber};
+
+  identify_ns.eui64 = 0x123456789abcdefa;
+  identify_ns.nguid[0] = 0x123456789abcdefa;
+  identify_ns.nguid[1] = 0x123456789abcdefa;
+
+  translator::StatusCode status =
+      translator::InquiryToScsi(scsi_cmd_, buffer, nvme_cmds_);
+  EXPECT_EQ(status, translator::StatusCode::kSuccess);
+
+  scsi::UnitSerialNumber result =
+      *reinterpret_cast<scsi::UnitSerialNumber*>(buffer);
+
+  EXPECT_EQ(result.peripheral_qualifier,
+            scsi::PeripheralQualifier::kPeripheralDeviceConnected);
+  EXPECT_EQ(result.peripheral_device_type,
+            scsi::PeripheralDeviceType::kDirectAccessBlock);
+  EXPECT_EQ(result.page_code, scsi::PageCode::kUnitSerialNumber);
+  EXPECT_EQ(result.page_length, 40);
+  char formatted_hex_string[41] = "1234_5678_9abc_defa_1234_5678_9abc_defa.";
+  uint8_t* product_serial_number =
+      reinterpret_cast<uint8_t*>(&buffer[sizeof(scsi::UnitSerialNumber)]);
+
+  for (int i = 0; i < 40; ++i) {
+    EXPECT_EQ(product_serial_number[i], formatted_hex_string[i]);
+  }
+}
+
+TEST_F(InquiryTest, TranslateUnitSerialNumberVpdNone) {
+  inquiry_cmd_ = scsi::InquiryCommand{
+      .evpd = 1, .page_code = scsi::PageCode::kUnitSerialNumber};
+
+  int8_t arr[20] = {'1', '2', '3', '4', '5', 'a', 'b', 'c', 'd', 'e',
+                    '1', '2', '3', '4', '5', 'a', 'b', 'c', 'd', 'e'};
+  memcpy(identify_ctrl.sn, arr, 20);
+  nvme_cmds_[1].nsid = 0xaaaaaaaa;
+
+  translator::StatusCode status =
+      translator::InquiryToScsi(scsi_cmd_, buffer, nvme_cmds_);
+  EXPECT_EQ(status, translator::StatusCode::kSuccess);
+
+  scsi::UnitSerialNumber result =
```

Comment:
could use ReadValue in some of the tests too

Reply:
Must, in fact--as written this is probably UB, as `buffer` is not guaranteed to be suitably aligned, whereas `ReadValue` has no alignment restrictions due to using an explicit memcpy.